stanford_soe_r25.module

```diff
@@ -0,0 +1,889 @@
+      }
+      // and if running commands event-get or billing-get, also
+      // return the raw XML to be modified.
+      if ($command == 'event-get' || $command == 'billing-get') {
+        $result['raw-xml'] = $output->data;
+      }
+    }
+  }
+  return $result;
+}
+
+/**
+ * Implementation of hook_mail for emailing reservations to staff
+ */
+function stanford_soe_r25_mail($key, &$message, $params) {
+  $message['subject'] = $params['subject'];
+  $message['body'] = $params['body'];
+}
+
+// function to parse the contents of a blackout date text area into an
+// array of start and end blackout dates
+function _stanford_soe_r25_parse_blackout_dates($inStr) {
+  $blackouts = array();
+  $blackout_text = trim($inStr);
+  if (!empty($blackout_text)) {
+    $tmp1 = explode("\n", $blackout_text);
+    foreach ($tmp1 as $tmp2) {
+      $blackout = trim($tmp2);
+      if (preg_match('(\d{4}-\d{2}-\d{2} - \d{4}-\d{2}-\d{2})', $blackout) === 1) {
+        $tmp3 = explode(" - ", $blackout);
+        $blackouts[] = array('start' => $tmp3[0], 'end' => $tmp3[1]);
+      }
+    }
+  }
+  return $blackouts;
+}
+
+// function to check if a given date is blacked out based on blackout dates
+// and current date.
+// $date is given as a UNIX timestamp
+function _stanford_soe_r25_date_blacked_out($date) {
+
+  // if an empty date is given, return false
+  if (empty($date)) {
+    return FALSE;
+  }
+
+  // if I'm currently in a blackout period, the requested date doesn't matter - we consider it blacked out.
+  // if I'm *not* currently in a blackout period, then the requested date has to be before the next blackout starts.
+  // if I'm currently past the last possible blackout, consider me blacked out so an admin will update the dates.
+  $blackouts = _stanford_soe_r25_parse_blackout_dates(variable_get('stanford_soe_r25_blackout_dates', ''));
+  $blacked_out = TRUE;
+
+  // first find out if we are currently in a blackout
+  // if not, find out when the next blackout starts
+  $cur = time();
+  $cur_blackout = TRUE;
+  $next_blackout = 0;
+  foreach ($blackouts as $blackout) {
+    if ($cur < strtotime($blackout['start'])) {
+      $cur_blackout = FALSE;
+      $next_blackout = strtotime($blackout['start']);
+      break;
+    }
+    else {
+      if ($cur <= strtotime($blackout['end'])) {
+        break;
+      }
+    }
+  }
+
+  // now see if the requested date is before the next blackout
+  if (!$cur_blackout && $date < $next_blackout) {
+    $blacked_out = FALSE;
+  }
+  return $blacked_out;
+}
+
+// this is a simple function to output room configurations for dev purposes.
+function _stanford_soe_r25_list_rooms() {
+  $rooms = _stanford_soe_r25_room_config_load();
+  $room_output = print_r($rooms, TRUE);
+  $output = 'rooms: <pre>' . $room_output . '</pre>';
+  return $output;
+}
+
+// check if the current user can confirm or cancel a reservation
+function _stanford_soe_r25_user_can_cancel_or_confirm($room_id, $event_id, $op) {
```

Comment:
refactor this function to be smaller/multiple function and easier to read. less indentations as well with return statements to replace if/else